src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx

```diff
@@ -0,0 +1,361 @@
+
+    this.state = {
+      allSelectedObjects: props.selectedSavedObjects,
+      // current workspace is the first option
+      workspaceOptions: [
+        ...(currentWorkspace ? [workspaceToOption(currentWorkspace, currentWorkspaceId)] : []),
+        ...targetWorkspacesOptions,
+      ],
+      targetWorkspaceOption: [],
+      isLoading: false,
+      isIncludeReferencesDeepChecked: true,
+      savedObjectTypeInfoMap,
+    };
+    this.isMounted = true;
+  }
+
+  componentWillUnmount() {
+    this.isMounted = false;
+  }
+
+  duplicateSavedObjects = async (savedObjects: SavedObjectWithMetadata[]) => {
+    this.setState({
+      isLoading: true,
+    });
+
+    const targetWorkspace = this.state.targetWorkspaceOption[0].key;
+
+    await this.props.onDuplicate(
+      savedObjects,
+      this.state.isIncludeReferencesDeepChecked,
+      targetWorkspace!
+    );
+
+    if (this.isMounted) {
+      this.setState({
+        isLoading: false,
+      });
+    }
+  };
+
+  onTargetWorkspaceChange = (targetWorkspaceOption: WorkspaceOption[]) => {
+    this.setState({
+      targetWorkspaceOption,
+    });
+  };
+
+  changeIncludeReferencesDeep = () => {
+    this.setState((state) => ({
+      isIncludeReferencesDeepChecked: !state.isIncludeReferencesDeepChecked,
+    }));
+  };
+
+  // Choose whether to copy a certain type or not.
+  changeIncludeSavedObjectType = (savedObjectType: string) => {
+    const { savedObjectTypeInfoMap } = this.state;
+    const savedObjectTypeInfo = savedObjectTypeInfoMap.get(savedObjectType);
+    if (savedObjectTypeInfo) {
+      const [count, checked] = savedObjectTypeInfo;
+      savedObjectTypeInfoMap.set(savedObjectType, [count, !checked]);
+      this.setState({ savedObjectTypeInfoMap });
+    }
+  };
+
+  isSavedObjectTypeIncluded = (savedObjectType: string) => {
+    const { savedObjectTypeInfoMap } = this.state;
+    const savedObjectTypeInfo = savedObjectTypeInfoMap.get(savedObjectType);
+    return savedObjectTypeInfo && savedObjectTypeInfo[1];
+  };
+
+  render() {
+    const {
+      workspaceOptions,
+      targetWorkspaceOption,
+      isIncludeReferencesDeepChecked,
+      allSelectedObjects,
+    } = this.state;
+    const { duplicateMode, onClose } = this.props;
+    const targetWorkspaceId = targetWorkspaceOption?.at(0)?.key;
+    let selectedObjects = allSelectedObjects;
+    if (duplicateMode === DuplicateMode.All) {
+      selectedObjects = selectedObjects.filter((item) => this.isSavedObjectTypeIncluded(item.type));
+    }
+    // If the target workspace is not selected, all saved objects will be retained.
+    // If the target workspace has been selected, filter out the saved objects that belongs to the workspace.
+    const includedSelectedObjects = selectedObjects.filter((item) =>
+      !!targetWorkspaceId && !!item.workspaces ? !item.workspaces.includes(targetWorkspaceId) : true
+    );
+
+    const ignoredSelectedObjectsLength = selectedObjects.length - includedSelectedObjects.length;
```

Comment:
Extract to a method of the component maybe? It is hard to understand what these lines are doing.

Reply:
Sure, have extracted the logic to a method.